Tools/paletteable sprite editor/js/lib/marvinj-0.9.js

```diff
@@ -0,0 +1,2665 @@
+		this.setAttribute("y", 0);
+		this.setAttribute("width", 0);
+		this.setAttribute("height", 0);
+	};
+	
+	Crop.prototype.process = function
+	(
+		imageIn, 
+		imageOut,
+		attributesOut,
+		mask, 
+		previewMode
+	)
+    {
+		var x = this.getAttribute("x");
+		var y = this.getAttribute("y");
+		var width = this.getAttribute("width");
+		var height = this.getAttribute("height");
+		
+		imageOut.setDimension(width, height);
+		
+		for(var i=x; i<x+width; i++){
+			for(var j=y; j<y+height; j++){
+				imageOut.setIntColor(i-x, j-y, imageIn.getIntColor(i, j));
+			}
+		}
+    };
+
+
+	function FloodfillSegmentation(){
+		MarvinAbstractImagePlugin.super(this);
+		this.load();
+	}
+
+	FloodfillSegmentation.prototype.load = function(){
+		this.setAttribute("returnType", "MarvinSegment");
+	};
+	
+	FloodfillSegmentation.prototype.process = function
+	(
+		imageIn, 
+		imageOut,
+		attributesOut,
+		mask, 
+		previewMode
+	)
+	{
+		if(attributesOut != null){
+			var returnType = this.getAttribute("returnType");
+			var fillBuffer = imageIn.clone();
+			var segments = this.floodfillSegmentation(imageIn, fillBuffer);
+			
+			switch(returnType){
+				case "MarvinSegment":
+					attributesOut.set("segments", segments);
+					break;
+				case "MarvinBlobSegment":
+					attributesOut.set("blobSegments", blobSegments(fillBuffer, segments));
+					break;
+			}
+		}
+	};
+	
+	FloodfillSegmentation.prototype.floodfillSegmentation = function(image, fillBuffer){
+		fillBuffer.clear(0xFF000000);
+		
+		var currentColor=1;
+		for(var y=0; y<image.getHeight(); y++){
+			for(var x=0; x<image.getWidth(); x++){
+				
+				var color = fillBuffer.getIntColor(x, y);
+				
+				if((color & 0x00FFFFFF) == 0 && image.getAlphaComponent(x, y) > 0){
+					var c = 0xFF000000 | (currentColor++);
+					Marvin.boundaryFill(image, fillBuffer, x, y, c);
+				}
+			}
+		}
+		
+		var segments = new Array(currentColor-1);
+		var seg;
+		for(var y=0; y<fillBuffer.getHeight(); y++){
+			for(var x=0; x<fillBuffer.getWidth(); x++){
+				var color = (fillBuffer.getIntColor(x, y) & 0x00FFFFFF);
+				
+				if(color != 0x00FFFFFF && color > 0){
+					
+					seg = segments[color-1];
+					
+					if(seg == null){
+						seg = new MarvinSegment();
+						segments[color-1] = seg;
+					}
+					
+					// x and width
+					if(seg.x1 == -1 || x < seg.x1)	{		seg.x1 = x;		}
+					if(seg.x2 == -1 || x > seg.x2)	{		seg.x2 = x;		}
+					seg.width = (seg.x2-seg.x1)+1;
+					
+					// y and height;
+					if(seg.y1 == -1 || y < seg.y1)	{		seg.y1 = y;		}
+					if(seg.y2 == -1 || y > seg.y2)	{		seg.y2 = y;		}
+					seg.height = (seg.y2-seg.y1)+1;
+					
+					seg.area++;
+				}
+			}
+		}
+		
+		return segments;
+	}
+	
+	FloodfillSegmentation.prototype.blobSegments = function(image, segments){
+		
+		var blobSegments = new Array(segments.length);
+		
+		var colorSegment;
+		var seg;
+		for(var i=0; i<segments.length; i++){
+			seg = segments[i];
+			colorSegment = 0xFF000000 + (i+1);
```

Comment:
Codacy found an issue: [The numeric literal '0xFF000000' will have at different value at runtime.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6395467)